File: puzzles/python3/there-is-no-spoon1/there_is_no_spoon1.py
from typing import List


if __name__ == "__main__":
    EMPTY = '.'
    NODE = '0'
    grid: List[str] = []

    width = int(input())  # the number of cells on the X axis
    height = int(input())  # the number of cells on the Y axis
    for _ in range(height):
        grid.append(input())

    for y in range(height):
        for x in range(width):
            if grid[y][x] == NODE:
                # node
                output = f"{x} {y} "

                # right neighbor
                neighbor = EMPTY
                for neighbor_x in range(x + 1, width):
                    neighbor = grid[y][neighbor_x]
                    if neighbor == NODE:
                        output += f"{neighbor_x} {y} "
                        break

                if neighbor == EMPTY:
                    output += "-1 -1 "

                # bottom neighbor
                neighbor = EMPTY
                for neighbor_y in range(y + 1, height):
                    neighbor = grid[neighbor_y][x]
                    if neighbor == NODE:
                        output += f"{x} {neighbor_y} "
                        break

                if neighbor == EMPTY:
                    output += "-1 -1 "

                print(output)

File: puzzles/kotlin/src/there-is-no-spoon1/there-is-no-spoon1.kt
import java.util.Scanner

fun main(args : Array<String>) {
  val empty = '.'
  val node = '0'

  val input = Scanner(System.`in`)
  val width = input.nextInt() // the number of cells on the X axis
  val height = input.nextInt() // the number of cells on the Y axis
  if (input.hasNextLine()) {
    input.nextLine()
  }
  val grid = ArrayList<String>(height)
  for (y in 0 until height) {
    val line = input.nextLine() // width characters, each either 0 or .
    grid.add(line)
  }

  for (y in 0 until height) {
    for (x in 0 until width) {
      if (grid[y][x] == node) {
        // node
        val output = StringBuilder("$x $y ")

        // right neighbor
        var neighbor = empty
        for (neighborX in x + 1 until width) {
          neighbor = grid[y][neighborX]
          if (neighbor == node) {
            output.append("$neighborX $y ")
            break
          }
        }

        if (neighbor == empty) {
          output.append("-1 -1 ")
        }

        // bottom neighbor
        neighbor = empty
        for (neighborY in y + 1 until height) {
          neighbor = grid[neighborY][x]
          if (neighbor == node) {
            output.append("$x $neighborY")
            break
          }
        }

        if (neighbor == empty) {
          output.append("-1 -1")
        }

        println(output)
      }
    }
  }
}
